MediaBrowser.Providers/Plugins/Tmdb/TmdbUtils.cs

```diff
@@ -201,5 +205,79 @@ public static string BuildParentalRating(string countryCode, string ratingValue)
+        public static string? RemoveInvalidFileCharacters(string? searchResultName)
+        {
+            if (searchResultName == null)
+            {
+                return null;
+            }
+
+            string formatted = _fileSystemInvalid.Replace(searchResultName, string.Empty);
+            return formatted;
+        }
+
+        /// <summary>
+        /// Identifies the nearest match to the search criteria. Addresses issues with search results from TMDB.
+        /// </summary>
+        /// <param name="name">The cleaned up name searched for.</param>
+        /// <param name="year">The year searched for.</param>
+        /// <param name="searchResults">The search results from a Movie or Tv search.</param>
+        /// <returns>The nearest match to the search criteria.</returns>
+        public static SearchMovieTvBase? GetBestMatch(string name, int year, IReadOnlyList<SearchMovieTvBase> searchResults)
+        {
+            List<SearchMovieTvBase> filtered = searchResults.ToList();
+
+            // partial matches can be returned before exact matches. Look for exact matches after removing invalid characters
+            var nearExactMatchFiltered = filtered.Where(s =>
+            {
+                string? resultName;
+                string? resultOriginalName;
+                switch (s)
+                {
+                    case SearchMovie searchMovie:
+                        resultName = searchMovie.Title;
+                        resultOriginalName = searchMovie.OriginalTitle;
+                        break;
+                    case SearchTv searchTv:
+                        resultName = searchTv.Name;
+                        resultOriginalName = searchTv.OriginalName;
+                        break;
+                    default:
+                        return false;
+                }
+
+                return string.Equals(RemoveInvalidFileCharacters(resultName), name, StringComparison.OrdinalIgnoreCase)
+                       || string.Equals(RemoveInvalidFileCharacters(resultOriginalName), name, StringComparison.OrdinalIgnoreCase);
+            }).ToList();
+
+            if (nearExactMatchFiltered.Any())
+            {
+                filtered = nearExactMatchFiltered;
+            }
+
+            // results of search can sometimes include poor matches (e.g. Hercules (2014) yielding Hercules (1997) as the top result
+            if (year > 0)
+            {
+                var yearFiltered = filtered.Where(s =>
+                {
+                    DateTime? searchResultYear = (s as SearchMovie)?.ReleaseDate
+                        ?? (s as SearchTv)?.FirstAirDate;
+                    return searchResultYear.HasValue
+                        && Math.Abs(searchResultYear.Value.Year - year) < MaxYearDifference;
+                }).ToList();
+                if (yearFiltered.Any())
+                {
+                    filtered = yearFiltered.ToList();
+                }
```

Comment:
Can just return the first matching value instead of putting everything into a list

```suggestion
                var yearFiltered = searchResults
                    .Where(s =>
                    {
                        DateTime? searchResultYear = (s as SearchMovie)?.ReleaseDate
                                                     ?? (s as SearchTv)?.FirstAirDate;
                        return searchResultYear.HasValue
                               && Math.Abs(searchResultYear.Value.Year - year) < MaxYearDifference;
                    })
                    .FirstOrDefault();

                if (yearFiltered is not null)
                {
                    return yearFiltered;
                }
```